active/0016-emqx-conf.md

```diff
@@ -0,0 +1,184 @@
+
+#### emqx.conf
+
+Emqx reads `emqx.conf` forconverting this hocon file into Erlang format at startup, and `emqx.conf` has only 2 lines by default.
+
+```erlang
+include "./data/emqx_cluster_override.conf"
+include "./data/emqx_local_override.conf"
+```
+
+- If the user wants to manually modify a node's configuration item before startup, it can be appended to the end of the `emqx.conf`, or use `include "./data/user_default.conf`, and for the same configuration, the later value will overwrite the earlier one.
+- If the user specifies to read environment variables for a configuration item, this value is read-only at runtime and will not be modified. In other words, the environment variables are always taken at the end of the `emqx.conf` file and has the highest priority.
+- Unlike the previous declaration of all configuration items displayed in `emqx.conf`, if an item uses a default value, it does not need to be shown in `emqx.conf`.  the default is only embedded in the code. Also, users can view all configuration items via the HTTP API (described later). This would have 2 benefits: 
+  - In subsequent version upgrades, adding/removing/updating configurations will not be overwritten by `emqx.conf`. 
+  - This allows the user to focus only on the configuration that has been modified. 
+
+#### emqx_cluster_override.conf
+
+- This file can only be modified via the API and manual modification of file directly is not supported.
+
+- When updating configurations that must require consistency across the cluster, they are persisted to this file.
+- The node will copy this file from the longest surviving core node before initializing the configuration, , this file will be added to initialize the configuration together.
+- When the node is updated, the configuration within the cluster is updated via cluster call and persisted to this file.
+- This file must be kept the same for all nodes, we will add an extra process to check the content of the file periodically, and alarm after 3 continuous differences are found.
+
+#### emqx_local_override.conf
+
+- This file can only be modified via the API and manual modification of file directly is not supported.
+
+- When the configuration of a specific node is updated via HTTP API, it will be persisted to this file.
+
+#### emqx_conf application
+Before, the initialization of the configuration file, cluster_call, was done through the `emqx_machine`, we will split this part of the functionality from the` emqx_machine` and make a new application `emqx_conf`
+
+The role of this application is to
+- Converting the configuration from hocon format to erlang sys.config format at initialization.
+- Manage hot updates and deletions of configurations within the cluster.
+
+The basic node configuration hot update remains in emqx, in other words: `emqx_config_handler`, remains unchanged, but other apps that want to update the configuration must call through the `emqx_conf`'s' API, which cannot call emqx API directly.
```

Comment:
This is a new feature for 5.0, so we are comparing it to 4.3, not 5.0 beta.1
Describing the changes on beta.1 will soon become stale.
So we should introduce `emqx_config_handler` as a new module.

It should be enough just to elaborate a bit on the layering of the APIs (line 64)